app/src/main/kotlin/com/wire/android/ui/home/settings/backup/BackupAndRestoreViewModel.kt

```diff
@@ -1,17 +1,272 @@
+                if (result.isEncrypted) {
+                    latestExtractedBackupRootPath = result.extractedFilesRootPath
+                    showPasswordDialog()
+                } else {
+                    importDatabase(result.extractedFilesRootPath)
+                }
+            }
+            is ExtractCompressedBackupFileResult.Failure -> {
+                state = state.copy(restoreFileValidation = RestoreFileValidation.IncompatibleBackup)
+                appLogger.e("Failed to extract backup files: ${result.error}")
+            }
+        }
+    }
+
+    private suspend fun importDatabase(extractedBackupRootPath: Path) {
+        state = state.copy(
+            restoreFileValidation = RestoreFileValidation.ValidNonEncryptedBackup,
+            backupRestoreProgress = BackupRestoreProgress.InProgress(0.75f)
+        )
+        when (importBackup(extractedBackupRootPath, null)) {
+            RestoreBackupResult.Success -> {
+                state = state.copy(backupRestoreProgress = BackupRestoreProgress.InProgress(0.75f))
+                delay(300)
+                state = state.copy(backupRestoreProgress = BackupRestoreProgress.Finished)
+            }
+            is RestoreBackupResult.Failure -> {
+                appLogger.e(
+                    "Error when restoring the db file. The format or version of the backup is not compatible with this " +
+                            "version of the app"
+                )
+                state = state.copy(
+                    restoreFileValidation = RestoreFileValidation.IncompatibleBackup,
+                    backupRestoreProgress = BackupRestoreProgress.Failed
+                )
+            }
+        }
+    }
+
+    @Suppress("MagicNumber")
+    fun restorePasswordProtectedBackup(restorePassword: TextFieldValue) = viewModelScope.launch {
+        state = state.copy(
+            backupRestoreProgress = BackupRestoreProgress.InProgress(0.50f),
+            restorePasswordValidation = PasswordValidation.NotVerified
+        )
+        delay(250)
+        val fileValidationState = state.restoreFileValidation
+        if (fileValidationState is RestoreFileValidation.PasswordRequired) {
+            when (val result = importBackup(latestExtractedBackupRootPath, restorePassword.text)) {
+                RestoreBackupResult.Success -> {
+                    state = state.copy(
+                        backupRestoreProgress = BackupRestoreProgress.Finished,
+                        restorePasswordValidation = PasswordValidation.Valid
+                    )
+                }
+                is RestoreBackupResult.Failure -> {
+                    mapBackupRestoreFailure(result.failure)
+                }
+            }
+        } else {
+            state = state.copy(backupRestoreProgress = BackupRestoreProgress.Failed)
+        }
+    }
+
+    private fun mapBackupRestoreFailure(failure: RestoreBackupResult.BackupRestoreFailure) = when (failure) {
```

Comment:
Could you move it to mapper ?

Reply:
in order to do that I would need to kind of pass the state to the constructor of the mapper and keep updating it from the mapper by passing the failure I do not think it is worth it since mapper will not be really generic and be dependent only on the state object of this viewodel